File: frontend/tests/wait-for-http.mjs
import { setTimeout as delay } from "node:timers/promises";

async function waitForHttp(url, timeoutMs = 60000) {
  const deadline = Date.now() + timeoutMs;
  let lastError;

  while (Date.now() < deadline) {
    try {
      const response = await fetch(url, { redirect: "manual" });
      if (response.status >= 200 && response.status < 500) {
        return;
      }
    } catch (error) {
      lastError = error;
    }

    await delay(1000);
  }

  throw new Error(`timed out waiting for ${url}: ${lastError?.message || lastError || "unknown error"}`);
}

const targetUrl = process.argv[2];

if (!targetUrl) {
  throw new Error("usage: node tests/wait-for-http.mjs <url>");
}

await waitForHttp(targetUrl);

File: backend/tests/wait_for_http.py
from __future__ import annotations

import sys
import time
from urllib.error import URLError
from urllib.request import urlopen


def main() -> int:
    url = sys.argv[1]
    deadline = time.time() + 60
    last_error = None

    while time.time() < deadline:
        try:
            with urlopen(url) as response:
                if 200 <= response.status < 500:
                    return 0
        except URLError as exc:
            last_error = exc
            time.sleep(1)

    raise SystemExit(f"timed out waiting for {url}: {last_error}")


if __name__ == "__main__":
    raise SystemExit(main())